spread/openstack.go

```diff
@@ -0,0 +1,640 @@
+	// simpler than in the google backend. Google checks for a)
+	// exact match, b) family match, c) term match, only (a), (c)
+	// are done here as there is no (b) in openstack. If multiple
+	// term matches are found the newest image is selected.
+	searchTerms := toTerms(imageName)
+	// First consider exact matches on name.
+	for _, image := range images {
+		// First consider exact matches on name.
+		if image.Name == imageName {
+			// return the image when it matchs exactly with the provided name
+			debugf("Name match: %#v matches %#v", imageName, image)
+			return &image, nil
+		}
+
+		// Otherwise use term matching.
+		imageTerms := toTerms(image.Name)
+		if containsTerms(imageTerms, searchTerms) {
+			debugf("Terms match: %#v matches %#v", imageName, image)
+			// Check if the creation date for the current image is after the previous selected one
+			currCreatedDate, err := time.Parse(time.RFC3339, image.Created)
+			// When the creation date is not set or it cannot be parsed, it is considered as created just now
+			if err != nil {
+				currCreatedDate = time.Time{}
+			}
+
+			// Save the image when either it is the first match or it is newer than the previous match
+			if lastImage.Id == "" || currCreatedDate.After(lastCreatedDate) {
+				debugf("Newer terms match found: %#v", image)
+				lastImage = image
+				lastCreatedDate = currCreatedDate
+			}
+		}
+	}
+
+	if lastImage.Id != "" {
+		return &lastImage, nil
+	}
+
+	return nil, &FatalError{fmt.Errorf("cannot find matching image for %s", imageName)}
+}
+
+func (p *openstackProvider) findAvailabilityZone() (*nova.AvailabilityZone, error) {
+	zones, err := p.computeClient.ListAvailabilityZones()
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve availability zones: %s", errorMsg(err))
+	}
+
+	if len(zones) == 0 {
+		return nil, &FatalError{errors.New("cannot find any availability zones")}
+	}
+	return &zones[0], nil
+}
+
+func (p *openstackProvider) findSecurityGroupNames(names []string) ([]nova.SecurityGroupName, error) {
+	var secGroupNames []nova.SecurityGroupName
+
+	secGroups, err := p.networkClient.ListSecurityGroupsV2()
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve security groups: %s", errorMsg(err))
+	}
+
+	if len(secGroups) == 0 {
+		return nil, &FatalError{errors.New("cannot find any groups")}
+	}
+
+	for _, name := range names {
+		found := false
+		for _, sg := range secGroups {
+			if name == sg.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, &FatalError{fmt.Errorf("cannot find valid group with name %s", name)}
+		}
+		secGroupNames = append(secGroupNames, nova.SecurityGroupName{Name: name})
+	}
+	return secGroupNames, nil
+}
+
+func (p *openstackProvider) waitServerCompleteBuilding(s *openstackServer, timeoutSeconds int) error {
+	// Wait until the server is actually running
+	start := time.Now()
+	for {
+		if time.Since(start) > time.Duration(timeoutSeconds)*time.Second {
```

Comment:
would it be a better idea to use channel timeouts to follow waiting pattern done in `googleProvider.waitOperation`?

Reply:
Yes, that would probably a nicer pattern here, 

Here is an (untested) diff, I wanted to write a proper unit test to ensure this works correctly but it's too late in the evening for it :/
[waitOpPattern.diff.txt](https://github.com/snapcore/spread/files/12921775/waitOpPattern.diff.txt)